### ISSUE TITLE
Why dynet cannot learn logistic regression well?

### ISSUE THREAD

**Issue by lihuiknight** [clab/dynet#1245]
I define my network based on dynet like below

```python
def network_lr(input, loss_arr, prob_arr, rnn=None):
    param_dict = input['param_dict']
    event_x_dim = input['event_x_dim']
    cur_event_feat = input['cur_event_feat']
    cur_label = input['cur_label']

    P = dynet.parameter

    x = dynet.vecInput(event_x_dim)
    x.set(cur_event_feat)
    y = dynet.scalarInput(cur_label)

    logit = P(param_dict['w_lr']) * x + P(param_dict['b_lr'])
    # logit = P(param_dict['w']) * P(param_dict['wx']) * x + P(param_dict['b'])
    prob = dynet.logistic(logit)
    loss = dynet.binary_log_loss(prob, y)
    loss_arr.append(loss)
    prob_arr.append(prob)


    return loss_arr, prob_arr


def init_flags():
    flags = tf.app.flags
    FLAGS = flags.FLAGS
    flags.DEFINE_float('learning_rate', 0.001, 'Initial learning rate.')
    flags.DEFINE_integer('display_step', 100, 'Display logs per step.')
    flags.DEFINE_integer('batch_size', 100, 'batch size')
    flags.DEFINE_integer('training_epochs', 10000, 'training epoch')

    # flags.DEFINE_integer('n_hidden_1', 32, '1st layer number of neurons')
    # flags.DEFINE_integer('n_hidden_2', 32, '1st layer number of neurons')
    flags.DEFINE_integer('num_embed', 200, 'layer number of neurons')
    flags.DEFINE_integer('x_dim', 2, 'feature dim')
    flags.DEFINE_integer('num_classes', 2, 'feature dim')

    return FLAGS

def opt_neural_network(x_train, x_test, y_train, y_test):

    assert FLAGS.batch_size<len(x_train), "batch_size should smaller than train sample size"

    param_dict = init_parameter(FLAGS.x_dim, FLAGS.num_embed)
    trainer = dynet.SimpleSGDTrainer(param_dict['model'])
    # trainer = dynet.AdamTrainer(param_dict['model'], alpha=FLAGS.learning_rate)
    event_x_dim = FLAGS.x_dim

    # Training cycle
    for epoch in range(0, FLAGS.training_epochs):

        dynet.renew_cg()
        prob_arr = []
        loss_arr = []
        for i in range(len(x_train)):

            input = {}
            input['param_dict'] = param_dict
            input['event_x_dim'] = event_x_dim
            input['cur_event_feat'] = x_train[i]
            input['cur_label'] = 1 if y_train[i] == '0 1' else 0

            # loss_arr, prob_arr = network_nn(input, loss_arr, prob_arr)
            loss_arr, prob_arr = network_lr(input, loss_arr, prob_arr)

        batch_loss = dynet.average(loss_arr)
        batch_loss.backward()
        trainer.update()


        if epoch % FLAGS.display_step == 0:
            prob_arr = [prob.value() for prob in prob_arr]
            ks, roc_auc, accuracy = mm.mdl_eval_binary(prob_arr, (np.array(y_train)[:,1]).tolist())
            train_ret = "[TRN]Epoch:%04d, loss=%.9f, ks=%.4f,roc_auc=%.4f" \
                        % (epoch, batch_loss.value(), ks, roc_auc)

            logger.info(train_ret)`
```
It seems that the program cannot event learn logistic regression well,

**Comment by pmichel31415** [comment]
Hi, can you provide a bit more context? What does the training behaviour look like? Also there seems to be pieces of your code missing, I see no creation of `ParameterCollection` for instance. It would help with debugging. Thanks

**Comment by lihuiknight** [comment]
@pmichel31415 , the full code as shown below. Firstly, I make a tony dataset, the label = 1 if x0+x1>1.2 else 0, then i define a logistic regression model to learn this function. Ideally, the learned AUC should be 0.999, i do the same work based on Tensorflow and obtain the perfect AUC, but dynet cannot do the same work(dynet auc=0.8), why?

------------------------------------------------------
-- my full code as below
`import dynet
import numpy as np
import logging
import tensorflow as tf
import sys
from sklearn import metrics


logging.basicConfig(filename='example.log',format='%(asctime)s-%(name)s-%(levelname)s-%(message)s: ', level=logging.DEBUG)
logging.getLogger().addHandler(logging.StreamHandler())
logger = logging.getLogger(__name__)


# create parameter collection
def init_parameter(x_dim, embed_dim):
    init = dynet.NormalInitializer(0, var=0.001)
    model = dynet.ParameterCollection()

    param_dict = {}
    #param for logistic regression
    param_dict['w_lr'] = model.add_parameters((1, x_dim), init)
    param_dict['b_lr'] = model.add_parameters((1), init)


    param_dict['model'] = model

    return param_dict

def network_lr(input, loss_arr, prob_arr, rnn=None):
    param_dict = input['param_dict']
    event_x_dim = input['event_x_dim']
    cur_event_feat = input['cur_event_feat']
    cur_label = input['cur_label']

    P = dynet.parameter

    x = dynet.vecInput(event_x_dim)
    x.set(cur_event_feat)
    y = dynet.scalarInput(cur_label)

    logit = P(param_dict['w_lr']) * x + P(param_dict['b_lr'])
    # logit = P(param_dict['w']) * P(param_dict['wx']) * x + P(param_dict['b'])
    prob = dynet.logistic(logit)
    loss = dynet.binary_log_loss(prob, y)
    loss_arr.append(loss)
    prob_arr.append(prob)


    return loss_arr, prob_arr


def init_flags():
    flags = tf.app.flags
    FLAGS = flags.FLAGS
    flags.DEFINE_float('learning_rate', 0.001, 'Initial learning rate.')
    flags.DEFINE_integer('display_step', 100, 'Display logs per step.')
    flags.DEFINE_integer('batch_size', 100, 'batch size')
    flags.DEFINE_integer('training_epochs', 10000, 'training epoch')

    flags.DEFINE_integer('num_embed', 200, 'layer number of neurons')
    flags.DEFINE_integer('x_dim', 2, 'feature dim')
    flags.DEFINE_integer('num_classes', 2, 'feature dim')

    return FLAGS


def opt_neural_network(x_train, x_test, y_train, y_test):

    assert FLAGS.batch_size<len(x_train), "batch_size should smaller than train sample size"

    param_dict = init_parameter(FLAGS.x_dim, FLAGS.num_embed)
    # trainer = dynet.SimpleSGDTrainer(param_dict['model'])
    trainer = dynet.AdamTrainer(param_dict['model'], alpha=FLAGS.learning_rate)
    event_x_dim = FLAGS.x_dim

    # Training cycle
    for epoch in range(0, FLAGS.training_epochs):

        dynet.renew_cg()
        prob_arr = []
        loss_arr = []
        for i in range(len(x_train)):

            input = {}
            input['param_dict'] = param_dict
            input['event_x_dim'] = event_x_dim
            input['cur_event_feat'] = x_train[i]
            input['cur_label'] = 1 if y_train[i] == '0 1' else 0

            loss_arr, prob_arr = network_lr(input, loss_arr, prob_arr)

        batch_loss = dynet.average(loss_arr)
        batch_loss.backward()
        trainer.update()


        if epoch % FLAGS.display_step == 0:
            prob_arr = [prob.value() for prob in prob_arr]
            roc_auc, accuracy = mdl_eval_binary(prob_arr, (np.array(y_train)[:,1]).tolist())
            train_ret = "[TRN]Epoch:%04d, loss=%.9f, roc_auc=%.4f" \
                        % (epoch, batch_loss.value(), roc_auc)

            logger.info(train_ret)

    logger.info("Optimization Finished!")


def mdl_eval_binary(pred, actual, bin_sz=100, is_print_gains=0, acc_cutoff=0.5):

    acc_matched = 0.0
    acc_total = 0.0

    prob_col = np.array(pred)
    y_col = np.array(actual)

    fpr, tpr, thresholds = metrics.roc_curve(y_col, prob_col)
    auc = metrics.auc(fpr, tpr)
    # print("auc=%.4f" % auc)
    roc_auc = auc

    for i in range(len(prob_col)):
        if prob_col[i]>=acc_cutoff and y_col[i] == 1:
            acc_matched = acc_matched + 1
        elif prob_col[i]<acc_cutoff and y_col[i] == 0:
            acc_matched = acc_matched + 1
        acc_total = acc_total + 1

    accuracy = acc_matched*1.0/acc_total

    return roc_auc, accuracy

def simulate_basic_operator_data(n_samples,x_dim=2, operator='xor'):
    import random as rd
    np.random.seed(0)
    # x_dim = 2
    x = []
    y = []

    for i in range(0, n_samples):
        x.append([rd.uniform(-1, 1) for _ in range(0, x_dim)])
        if operator == 'xor':
            y.append([0, 1] if x[i][0]<0.4 and x[i][1]>0.6 else [1, 0])
        elif operator == 'multiple':
            y.append([0, 1] if x[i][0]*x[i][1] > 0.6 else [1, 0])
        elif operator == 'divide':
            y.append([0, 1] if x[i][1]>0 and x[i][0]/x[i][1] > 1.5 else [1, 0])
        elif operator == 'add':
            y.append([0, 1] if x[i][0] + x[i][1] > 1.2 else [1, 0])

    print('event_rate=%.4f' % (np.sum(np.array(y)[:,1])*1.0/n_samples) )
    cutoff = int(n_samples*0.8)
    return x[0:cutoff], x[cutoff:],y[0:cutoff], y[cutoff:]

if __name__ == '__main__':
    FLAGS = init_flags()
    logger.info('Load data ...')
    n_samples = 1000
    x_train, x_test, y_train, y_test = simulate_basic_operator_data(n_samples=n_samples, x_dim=2, operator='add')

    logger.info('Training ...')
    opt_neural_network(x_train, x_test, y_train, y_test)`

**Comment by FilippoC** [comment]
Did you try updating the learning_rate after each epoch?

**Comment by pmichel31415** [comment]
Hmm OK so I took a look and tried to run your code but there are several things that make it hard to conform your result:

1. The experiment is not deterministic: the dataset is random with a different seed in each run. Also you are sampling inputs from `[-1, 1]^x_dim` and given that your decision boundary  is `x+y =0.5` on average the class distribution should be something like 28%/72%. On some runs I got 0.99 AUC after one update, sometimes I got 0.4 at the end of training

2. Your initial learning rate is 0.1 for SGD (dynet's default) which might be too big

3. What @FilippoC  said
